dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java

```diff
@@ -478,27 +483,232 @@ public Object put(String key, Object value) {
+        }
+        @Override
+        public final boolean hasNext() {
+            if (backedIterator.hasNext()) {
+                return true;
+            }
+            if (next == TOTAL) {
+                return false;
+            }
+
+            for (int i = current+1; i < TOTAL; i++) {
+                if (propertyValues[i] != NOT_FOUND) {
+                    next = i;
+                    return true;
+                }
             }
+            next = TOTAL;
+            return false;
+        }
+        @Override
+        public final T next() {
+            if (hasNext()) {
+                if (next == -1) {
+                    removeAllowed = true;
+                    return backedIterator.next();
+                } else {
+                    current = next;
+                    removeAllowed = true;
+                    return getNextFromProperties(next);
+                }
+            }
+            removeAllowed = false;
+            throw new NoSuchElementException();
+        }
+        @Override
+        public final void remove() {
+            if (!removeAllowed) {
+                throw new IllegalStateException();
+            }
+            if (current == -1) {
+                backedIterator.remove();
+            } else {
+                propertyValues[current] = NOT_FOUND;
+            }
+            removeAllowed = false;
+        }
+        abstract T getNextFromProperties(int index);
+    }
+    
+    final class KeyIterator extends MessageIterator<String> {
+        KeyIterator() {
+            super(backedKeySet().iterator());
+        }
+
+        @Override
+        String getNextFromProperties(int index) {
+            return propertyNames[index];
+        }
+        
+    }
+
+    final class EntryIterator extends MessageIterator<Map.Entry<String, Object>> {
+        EntryIterator() {
+            super(backedEntrySet().iterator());
         }
 
-        return myKeys;
+        @Override
+        Map.Entry<String, Object> getNextFromProperties(int index) {
+            return new AbstractMap.SimpleEntry<String, Object>(propertyNames[index], propertyValues[index]);
+        }
+        
     }
     
-    @Override
-    public Set<Map.Entry<String,Object>> entrySet() {
-        HashSet<Map.Entry<String,Object>> myEntrySet = new HashSet<Map.Entry<String,Object>>(super.entrySet());
-        for (int i = 0; i < TOTAL; i++) {
-            if (propertyValues[i] != NOT_FOUND) {
-                myEntrySet.add(new AbstractMap.SimpleEntry<String,Object>(propertyNames[i], propertyValues[i]));
+    final class ValuesIterator extends MessageIterator<Object> {
+        ValuesIterator() {
+            super(backedValues().iterator());
+        }
+
+        @Override
+        Object getNextFromProperties(int index) {
+            return propertyValues[index];
+        }
+        
+    }
+    
+    final class KeySet extends AbstractSet<String> {
+        public final int size() {
+            return this.size();
```

Comment:
This will create an infinite loop.  clear() has the same issue.  You need to change to MessageImpl.this.size() and MessageImpl.this.clear();

Reply:
Similar issue in the other impls as well.